File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/overlay/overlay/NMarker.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.overlay

import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.Align
import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asBoolean
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asFloat
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asAlign
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.toMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.toMessageableString
import dev.note11.flutter_naver_map.flutter_naver_map.model.NPoint
import dev.note11.flutter_naver_map.flutter_naver_map.model.NSize
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.NOverlayCaption
import dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.NOverlayImage
import dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil.dpToPx
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil.pxToDp

internal data class NMarker(
    override val info: NOverlayInfo,
    val position: LatLng,
    val icon: NOverlayImage?,
    val iconTintColor: Int,
    val alpha: Float,
    val angle: Float,
    val anchor: NPoint,
    val size: NSize,
    val caption: NOverlayCaption?,
    val subCaption: NOverlayCaption?,
    val captionAligns: List<Align>,
    val captionDpOffset: Double,
    val isCaptionPerspectiveEnabled: Boolean,
    val isIconPerspectiveEnabled: Boolean,
    val isFlat: Boolean,
    val isForceShowCaption: Boolean,
    val isForceShowIcon: Boolean,
    val isHideCollidedCaptions: Boolean,
    val isHideCollidedMarkers: Boolean,
    val isHideCollidedSymbols: Boolean,
) : AddableOverlay<Marker> {

    override fun createMapOverlay(): Marker = Marker(position).also { m ->
        icon?.applyToOverlay(m::setIcon)
        m.iconTintColor = iconTintColor
        m.alpha = alpha
        m.angle = angle
        m.anchor = anchor.toPointF()
        size.useAsPixelSize(m::setWidth, m::setHeight)
        caption?.useWithFunctions(
            textFunc = m::setCaptionText,
            textSizeFunc = m::setCaptionTextSize,
            colorFunc = m::setCaptionColor,
            haloColorFunc = m::setCaptionHaloColor,
            minZoomFunc = m::setCaptionMinZoom,
            maxZoomFunc = m::setCaptionMaxZoom,
            requestWidthFunc = m::setCaptionRequestedWidth,
        )
        subCaption?.useWithFunctions(
            textFunc = m::setSubCaptionText,
            textSizeFunc = m::setSubCaptionTextSize,
            colorFunc = m::setSubCaptionColor,
            haloColorFunc = m::setSubCaptionHaloColor,
            minZoomFunc = m::setSubCaptionMinZoom,
            maxZoomFunc = m::setSubCaptionMaxZoom,
            requestWidthFunc = m::setSubCaptionRequestedWidth,
        )
        m.setCaptionAligns(*captionAligns.toTypedArray())
        m.captionOffset = dpToPx(captionDpOffset)
        m.isCaptionPerspectiveEnabled = isCaptionPerspectiveEnabled
        m.isIconPerspectiveEnabled = isIconPerspectiveEnabled
        m.isFlat = isFlat
        m.isForceShowCaption = isForceShowCaption
        m.isForceShowIcon = isForceShowIcon
        m.isHideCollidedCaptions = isHideCollidedCaptions
        m.isHideCollidedMarkers = isHideCollidedMarkers
        m.isHideCollidedSymbols = isHideCollidedSymbols
    }

    override fun toMap(): Map<String, Any?> = mapOf(
        infoName to info.toMap(),
        iconName to null,
        positionName to position.toMap(),
        iconTintColorName to iconTintColor,
        alphaName to alpha,
        angleName to angle,
        anchorName to anchor.toMap(),
        sizeName to size.toMap(),
        captionName to caption?.toMap(),
        subCaptionName to subCaption?.toMap(),
        captionAlignsName to captionAligns.map { it.toMessageableString() },
        captionOffsetName to captionDpOffset,
        isCaptionPerspectiveEnabledName to isCaptionPerspectiveEnabled,
        isIconPerspectiveEnabledName to isIconPerspectiveEnabled,
        isFlatName to isFlat,
        isForceShowCaptionName to isForceShowCaption,
        isForceShowIconName to isForceShowIcon,
        isHideCollidedCaptionsName to isHideCollidedCaptions,
        isHideCollidedMarkersName to isHideCollidedMarkers,
        isHideCollidedSymbolsName to isHideCollidedSymbols,
    )

    companion object {
        fun fromMap(rawMap: Any): NMarker = rawMap.asMap().let {
            NMarker(
                info = NOverlayInfo.fromMap(it[infoName]!!),
                position = it[positionName]!!.asLatLng(),
                icon = it[iconName]?.let(NOverlayImage.Companion::fromMap),
                iconTintColor = it[iconTintColorName]!!.asInt(),
                alpha = it[alphaName]!!.asFloat(),
                angle = it[angleName]!!.asFloat(),
                anchor = NPoint.fromMap(it[anchorName]!!),
                size = NSize.fromMap(it[sizeName]!!),
                caption = it[captionName]?.let(NOverlayCaption.Companion::fromMap),
                subCaption = it[subCaptionName]?.let(NOverlayCaption.Companion::fromMap),
                captionAligns = it[captionAlignsName]!!.asList { raw -> raw.asAlign() },
                captionDpOffset = it[captionOffsetName]!!.asDouble(),
                isCaptionPerspectiveEnabled = it[isCaptionPerspectiveEnabledName]!!.asBoolean(),
                isIconPerspectiveEnabled = it[isIconPerspectiveEnabledName]!!.asBoolean(),
                isFlat = it[isFlatName]!!.asBoolean(),
                isForceShowCaption = it[isForceShowCaptionName]!!.asBoolean(),
                isForceShowIcon = it[isForceShowIconName]!!.asBoolean(),
                isHideCollidedCaptions = it[isHideCollidedCaptionsName]!!.asBoolean(),
                isHideCollidedMarkers = it[isHideCollidedMarkersName]!!.asBoolean(),
                isHideCollidedSymbols = it[isHideCollidedSymbolsName]!!.asBoolean(),
            )
        }

        fun fromMarker(marker: Overlay, id: String): NMarker = (marker as Marker).run {
            NMarker(
                info = NOverlayInfo(NOverlayType.MARKER, id),
                position = position,
                icon = null,
                iconTintColor = iconTintColor,
                alpha = alpha,
                angle = angle,
                anchor = NPoint.fromPointF(anchor),
                size = NSize.fromPixelSize(width, height),
                caption = NOverlayCaption(
                    text = captionText,
                    textSize = captionTextSize,
                    color = captionColor,
                    haloColor = captionHaloColor,
                    minZoom = captionMinZoom,
                    maxZoom = captionMaxZoom,
                    requestWidth = pxToDp(captionRequestedWidth),
                ),
                subCaption = NOverlayCaption(
                    text = subCaptionText,
                    textSize = subCaptionTextSize,
                    color = subCaptionColor,
                    haloColor = subCaptionHaloColor,
                    minZoom = subCaptionMinZoom,
                    maxZoom = subCaptionMaxZoom,
                    requestWidth = pxToDp(subCaptionRequestedWidth),
                ),
                captionAligns = captionAligns.toList(),
                captionDpOffset = pxToDp(captionOffset),
                isCaptionPerspectiveEnabled = isCaptionPerspectiveEnabled,
                isIconPerspectiveEnabled = isIconPerspectiveEnabled,
                isFlat = isFlat,
                isForceShowCaption = isForceShowCaption,
                isForceShowIcon = isForceShowIcon,
                isHideCollidedCaptions = isHideCollidedCaptions,
                isHideCollidedMarkers = isHideCollidedMarkers,
                isHideCollidedSymbols = isHideCollidedSymbols,
            )
        }

        /*
            --- Messaging Name Define ---
        */

        private const val infoName = "info"
        const val hasOpenInfoWindowName = "hasOpenInfoWindow"
        const val openInfoWindowName = "openInfoWindow"
        const val positionName = "position"
        const val iconName = "icon"
        const val iconTintColorName = "iconTintColor"
        const val alphaName = "alpha"
        const val angleName = "angle"
        const val anchorName = "anchor"
        const val sizeName = "size"
        const val captionName = "caption"
        const val subCaptionName = "subCaption"
        const val captionAlignsName = "captionAligns"
        const val captionOffsetName = "captionOffset"
        const val isCaptionPerspectiveEnabledName = "isCaptionPerspectiveEnabled"
        const val isIconPerspectiveEnabledName = "isIconPerspectiveEnabled"
        const val isFlatName = "isFlat"
        const val isForceShowCaptionName = "isForceShowCaption"
        const val isForceShowIconName = "isForceShowIcon"
        const val isHideCollidedCaptionsName = "isHideCollidedCaptions"
        const val isHideCollidedMarkersName = "isHideCollidedMarkers"
        const val isHideCollidedSymbolsName = "isHideCollidedSymbols"
    }
}

File: ios/Classes/model/map/overlay/overlay/NMarker.swift
import NMapsMap

internal struct NMarker: AddableOverlay {
    typealias OverlayType = NMFMarker
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let position: NMGLatLng
    let icon: NOverlayImage?
    let iconTintColor: UIColor
    let alpha: CGFloat
    let angle: CGFloat
    let anchor: NPoint
    let size: NSize
    let caption: NOverlayCaption?
    let subCaption: NOverlayCaption?
    let captionAligns: Array<NMFAlignType>
    let captionOffset: CGFloat
    let isCaptionPerspectiveEnabled: Bool
    let isIconPerspectiveEnabled: Bool
    let isFlat: Bool
    let isForceShowCaption: Bool
    let isForceShowIcon: Bool
    let isHideCollidedCaptions: Bool
    let isHideCollidedMarkers: Bool
    let isHideCollidedSymbols: Bool

    func createMapOverlay() -> OverlayType {
        let marker = NMFMarker()
        return applyAtRawOverlay(marker)
    }
    
    func applyAtRawOverlay(_ marker: NMFMarker) -> NMFMarker {
        marker.position = position
        if let icon = icon {
            marker.iconImage = icon.overlayImage
        }
        marker.iconTintColor = iconTintColor
        marker.alpha = alpha
        marker.angle = angle
        marker.anchor = anchor.cgPoint
        marker.width = size.width
        marker.height = size.height
        
        if let caption = caption {
            marker.captionText = caption.text
            marker.captionTextSize = caption.textSize
            marker.captionColor = caption.color
            marker.captionHaloColor = caption.haloColor
            marker.captionMinZoom = caption.minZoom
            marker.captionMaxZoom = caption.maxZoom
            marker.captionRequestedWidth = caption.requestWidth
        }
        if let subCaption = subCaption {
            marker.subCaptionText = subCaption.text
            marker.subCaptionTextSize = subCaption.textSize
            marker.subCaptionColor = subCaption.color
            marker.subCaptionHaloColor = subCaption.haloColor
            marker.subCaptionMinZoom = subCaption.minZoom
            marker.subCaptionMaxZoom = subCaption.maxZoom
            marker.subCaptionRequestedWidth = subCaption.requestWidth
        }
        marker.captionAligns = captionAligns
        marker.captionOffset = captionOffset
        marker.captionPerspectiveEnabled = isCaptionPerspectiveEnabled
        marker.iconPerspectiveEnabled = isIconPerspectiveEnabled
        marker.isFlat = isFlat
        marker.isForceShowCaption = isForceShowCaption
        marker.isForceShowIcon = isForceShowIcon
        marker.isHideCollidedCaptions = isHideCollidedCaptions
        marker.isHideCollidedMarkers = isHideCollidedMarkers
        marker.isHideCollidedSymbols = isHideCollidedSymbols
        return marker
    }

    static func fromMessageable(_ v: Any) -> NMarker {
        let d = asDict(v)
        return NMarker(
                info: NOverlayInfo.fromMessageable(d[NMarker.infoName]!),
                position: asLatLng(d[NMarker.positionName]!),
                icon: castOrNull(d[NMarker.iconName], caster: NOverlayImage.fromMessageable),
                iconTintColor: asUIColor(d[NMarker.iconTintColorName]!),
                alpha: asCGFloat(d[NMarker.alphaName]!),
                angle: asCGFloat(d[NMarker.angleName]!),
                anchor: NPoint.fromMessageable(d[NMarker.anchorName]!),
                size: NSize.fromMessageable(d[NMarker.sizeName]!),
                caption: castOrNull(d[NMarker.captionName], caster: NOverlayCaption.fromMessageable),
                subCaption: castOrNull(d[NMarker.subCaptionName], caster: NOverlayCaption.fromMessageable),
                captionAligns: asArr(d[NMarker.captionAlignsName]!, elementCaster: asAlign),
                captionOffset: asCGFloat(d[NMarker.captionOffsetName]!),
                isCaptionPerspectiveEnabled: asBool(d[NMarker.isCaptionPerspectiveEnabledName]!),
                isIconPerspectiveEnabled: asBool(d[NMarker.isIconPerspectiveEnabledName]!),
                isFlat: asBool(d[NMarker.isFlatName]!),
                isForceShowCaption: asBool(d[NMarker.isForceShowCaptionName]!),
                isForceShowIcon: asBool(d[NMarker.isForceShowIconName]!),
                isHideCollidedCaptions: asBool(d[NMarker.isHideCollidedCaptionsName]!),
                isHideCollidedMarkers: asBool(d[NMarker.isHideCollidedMarkersName]!),
                isHideCollidedSymbols: asBool(d[NMarker.isHideCollidedSymbolsName]!)
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let hasOpenInfoWindowName = "hasOpenInfoWindow"
    static let openInfoWindowName = "openInfoWindow"
    static let positionName = "position"
    static let iconName = "icon"
    static let iconTintColorName = "iconTintColor"
    static let alphaName = "alpha"
    static let angleName = "angle"
    static let anchorName = "anchor"
    static let sizeName = "size"
    static let captionName = "caption"
    static let subCaptionName = "subCaption"
    static let captionAlignsName = "captionAligns"
    static let captionOffsetName = "captionOffset"
    static let isCaptionPerspectiveEnabledName = "isCaptionPerspectiveEnabled"
    static let isIconPerspectiveEnabledName = "isIconPerspectiveEnabled"
    static let isFlatName = "isFlat"
    static let isForceShowCaptionName = "isForceShowCaption"
    static let isForceShowIconName = "isForceShowIcon"
    static let isHideCollidedCaptionsName = "isHideCollidedCaptions"
    static let isHideCollidedMarkersName = "isHideCollidedMarkers"
    static let isHideCollidedSymbolsName = "isHideCollidedSymbols"
}
